### DISCUSSION TITLE
Game Development - SDL / DragonRuby-GTK

### DISCUSSION THREAD

**Discussion by ylluminate** [vlang/v#11609] [Ideas]
As @medvednikov mentioned, [SDL](https://en.wikipedia.org/wiki/Ryan_C._Gordon) library availability would be useful for V. (Note, [as per Wikipedia](https://en.wikipedia.org/wiki/Simple_DirectMedia_Layer): "A common misconception is that SDL [Simple DirectMedia Library] is a game engine, but this is not true. However, the library is suited to building games directly, or is usable indirectly by engines built on top of it.")

Since SDL is not specifically a game dev toolkit, let's look more closely at SDL's creator's sexiest new baby: DragonRuby-GTK

Since SDL creator (Ryan C. Gordon ) has now partnered up with the RubyMotion owner to make the DragonRuby Game Toolkit this might be the best place to look for inspiration. This can be reviewed here:
• https://dragonruby.itch.io/
• https://dragonruby.itch.io/dragonruby-gtk
• [announcement](http://www.rubymotion.com/news/2019/03/01/the-sleeping-dragon-has-awoken.html)
• Good marketing idea: [Build a game to learn how to program dammit! I'm here to help. Getting started instructions enclosed:
](https://www.reddit.com/r/learnprogramming/comments/clbm3t/build_a_game_to_learn_how_to_program_dammit_im/)

The examples on the main page are great:
• [Flappy Dragon](https://dragonruby.itch.io/flappydragon)
• [Basic Gorillas](https://dragonruby.itch.io/basicgorillas)
• [GTK Workshop Assets](https://dragonruby.itch.io/dragonruby-gtk-workshop)

Maybe @nsauzede's effort can be used as a foundation or core for this effort: https://github.com/nsauzede/vsdl2

**Comment by gslicer** [comment]
My dream is that we do not need to differentiate between 2D/3D app development - it should be feasible to render everything using OpenGL/Vulkan whatever - imagine your Windowed apps having smoothly animated perfectly rendered vector-based resolution independent widgets - imagine the possibilities.

See this interesting Project: [NanoGUI](https://github.com/wjakob/nanogui)

**Comment by dumblob** [comment]
@gslicer that's a nice dream, but I doubt it's practical due to high energy inefficiency. My take is to have them separate under the condition, that 2D API can be used seamlessly also in 3D (as in [Nuklear](https://github.com/vurtun/nuklear ) or even better [Quarks](https://gist.github.com/vurtun/c5b0374c27d2f5e9905bfbe7431d9dc0 )), **but not vice versa** (3D can per its definition not be that energy efficient). This allows applications do most of their stuff using 2D and only when there is enough power, use also 3D.

**Comment by nsauzede** [comment]
@gslicer did you note that SDL2 (wrapped by the nsauzede.vsdl2 module on vpm) does support just that : graphics (even 2D) are handled by 3D textures, because it interfaces with OpenGL with HW accel
so you mix 2D/3D naturally
I think there are some UI SDL2 plugins for widgets -- not sure, must check
I can have a look to create a V SDL2 3D example...

**Comment by dumblob** [comment]
>I think there are some UI SDL2 plugins for widgets -- not sure, must check

As I pointed out above, [Nuklear](https://github.com/vurtun/nuklear ), [Quarks](https://gist.github.com/vurtun/c5b0374c27d2f5e9905bfbe7431d9dc0 ), [ImGUI](https://github.com/ocornut/imgui ) all have backends for SDL and thus do exactly what you write.

**Comment by nsauzede** [comment]
> Since SDL creator (Ryan C. Gordon ) has now partnered up with the RubyMotion owner to make the DragonRuby Game Toolkit this might be the best place to look for inspiration. This can be reviewed here:
> • https://dragonruby.itch.io/

Just for the record, Ryan C. Gordon is not the SDL creator, but [Sam Lantinga (slouken)](https://en.wikipedia.org/wiki/Sam_Lantinga) is.
However, [Ryan](https://en.wikipedia.org/wiki/Ryan_C._Gordon) seems to be largely involved (see my comment below) in the SDL project, hence the slight confusion..

**Comment by nsauzede** [comment]
> @gslicer did you note that SDL2 (wrapped by the nsauzede.vsdl2 module on vpm) does support just that : graphics (even 2D) are handled by 3D textures, because it interfaces with OpenGL with HW accel
> so you mix 2D/3D naturally

=> reply to self : this is untrue
Apparently, Sam and Ryan are still wondering if/how to integrate some long-standing SDL2 patch to enable 2D/3D mixing :
https://bugzilla.libsdl.org/show_bug.cgi?id=1734#c17

> I think there are some UI SDL2 plugins for widgets -- not sure, must check
> I can have a look to create a V SDL2 3D example...
